File: src/jake2/sys/Timer.java
/*
 * Timer.java
 * Copyright (C) 2005
 */
package jake2.sys;

import jake2.client.Context;
import jake2.qcommon.Command;


public abstract class Timer {

	abstract public long currentTimeMillis();
	static Timer t;
	
	static {
		try {
			t = new NanoTimer();
		} catch (Throwable e) {
			try {
				t = new HighPrecisionTimer();
			} catch (Throwable e1) {
				t = new StandardTimer();
			}
		}
		Command.Println("using " + t.getClass().getName());
	}
	
	public static int Milliseconds() {
		return Context.curtime = (int)(t.currentTimeMillis());
	}
}

File: src/jake2/qcommon/Timer.kt
/*
 * Timer.java
 * Copyright (C) 2005
 */
package jake2.qcommon

import jake2.client.Context


class Timer private constructor()/* Singleton. */ {

    private val base = System.nanoTime()

    private fun currentTimeMillis(): Long {
        val time = System.nanoTime()
        var delta = time - base
        if (delta < 0) {
            delta += java.lang.Long.MAX_VALUE + 1
        }
        return (delta * 0.000001).toLong()
    }

    companion object {

        private val t = Timer()

        @JvmStatic fun Milliseconds(): Int {
            Context.curtime = t.currentTimeMillis().toInt()
            return Context.curtime
        }
    }
}
